indexer/services/ender/src/lib/kafka-publisher.ts

```diff
@@ -43,49 +55,122 @@ export class KafkaPublisher {
   }
 
   public addEvent(event: ConsolidatedKafkaEvent) {
-    switch (event.topic) {
+    this.getMessages(event.topic)!.push(event.message);
+  }
+
+  /**
+   * Helper function to get messages for a given topic.
+   *
+   * @param kafkaTopic
+   * @private
+   */
+  private getMessages(kafkaTopic: KafkaTopics): Message[] | undefined {
+    switch (kafkaTopic) {
       case KafkaTopics.TO_WEBSOCKETS_SUBACCOUNTS:
-        this.subaccountMessages.push(event.message);
-        break;
+        return this.subaccountMessages;
       case KafkaTopics.TO_WEBSOCKETS_TRADES:
-        this.tradeMessages.push(event.message);
-        break;
+        return this.tradeMessages;
       case KafkaTopics.TO_WEBSOCKETS_MARKETS:
-        this.marketMessages.push(event.message);
-        break;
+        return this.marketMessages;
       case KafkaTopics.TO_WEBSOCKETS_CANDLES:
-        this.candleMessages.push(event.message);
-        break;
+        return this.candleMessages;
       case KafkaTopics.TO_VULCAN:
-        this.vulcanMessages.push(event.message);
-        break;
+        return this.vulcanMessages;
       default:
         throw new Error('Invalid Topic');
     }
   }
 
   /**
-   * Sort trade events by block height, transaction index, and event index in ascending order,
-   * where the first trade event should be the earliest trade in the block.
+   * Sort subaccountMessages that represent fills by block height, transaction index,
+   * and event index in ascending order per order id. Only keep the subaccount message if
+   * it represents the last fill event per order id, but make sure the subaccount message
+   * contents contains all individual fills from that block.
+   *
+   * Due to separate handlers for order fills, we can be sure that if a message is annotated
+   * with a fill, it should only contain data about a single fill / order and not transfers.
    */
-  public sortTradeEvents() {
-    this.tradeMessages = this.tradeMessages.sort(
-      (a: SingleTradeMessage, b: SingleTradeMessage) => {
-        if (Big(a.blockHeight).lt(b.blockHeight)) {
-          return -1;
-        } else if (Big(a.blockHeight).gt(b.blockHeight)) {
-          return 1;
+  public aggregateFillEventsForSubaccountMessages() {
```

Comment:
nit: I think it still makes sense to add the TODO, to genericize so we can do a similar thing for positions / create a linear ticket.

Reply:
done